beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -2616,21 +2591,29 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
                 .map_err(BeaconChainError::from)?;
         }
 
-        let mut fork_choice = self.fork_choice.write();
+        let mut fork_choice = self.canonical_head.fork_choice_write_lock();
 
         // Do not import a block that doesn't descend from the finalized root.
-        let signed_block =
-            check_block_is_finalized_descendant::<T, _>(signed_block, &fork_choice, &self.store)?;
-        let (block, block_signature) = signed_block.clone().deconstruct();
+        check_block_is_finalized_descendant(self, &fork_choice, &signed_block)?;
 
-        // compare the existing finalized checkpoint with the incoming block's finalized checkpoint
-        let old_finalized_checkpoint = fork_choice.finalized_checkpoint();
+        // Note: we're using the finalized checkpoint from the head state, rather than fork
+        // choice.
+        //
+        // We are doing this to ensure that we detect changes in finalization. It's possible
+        // that fork choice has already been updated to the finalized checkpoint in the block
+        // we're importing.
+        let current_head_finalized_checkpoint =
+            self.canonical_head.cached_head().finalized_checkpoint();
+        // Compare the existing finalized checkpoint with the incoming block's finalized checkpoint.
         let new_finalized_checkpoint = state.finalized_checkpoint();
 
```

Comment:
The variables `current_head_finalized_checkpoint` and `new_finalized_checkpoint` are only used inside the `if let Some(wss_checkpoint) = self.config.weak_subjectivity_checkpoint {` statement below so they should probably be moved inside that if block to avoid unnecessary `clone`ing when not using weak subjectivity sync.

Reply:
Nice tidy up! Addressed in 65c1da107.